data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,454 @@
+---
+id: metaprogramming
+title: Preprocessors and PPXs
+description: |
+  An introduction to metaprogramming in OCaml, including preprocessors, PPX
+  extensions and the `ppxlib` library.
+category: "guides"
+date: 2022-07-25T21:07:30-00:00
+---
+
+
+# Preprocessors
+
+Preprocessors are programs meant to be called at compile time, so that they
+alter the program's source code before the actual compilation. They can be very
+useful for several things, such as the inclusion of a file, conditional
+compilation, or use of macros; but they're also useful in OCaml to generate a function from a
+type definition.
+
+To start with an example, here is how the following source code would be altered
+by [this
+preprocessor](https://github.com/ocaml-ppx/ppxlib/tree/main/examples/simple-extension-rewriter):
+
+
+```ocaml
+Printf.printf "This program has been compiled on %s" [%get_env "OSTYPE"]
+```
+
+would become:
+
+```ocaml
+Printf.printf "This program has been compiled on %s" "linux-gnu"
+```
+
+
+At compile time, the preprocessor would replace `[%get_env "OSTYPE"]` by a
+string with the content of the `OSTYPE` environment variable. Note that this
+happens at _compile time_, so at _run time_ the value of the `OSTYPE` variable
+would have no effect.
+
+## Source Preprocessors and PPXs
+
+In OCaml, the preprocessors are written by the community, so there is no single
+general-purpose preprocessor officially supported, like there is in C, for example.
+
+OCaml supports the execution of two kinds of preprocessors: one that works on the
+source level, and the other that works on the [AST
+level](#ocamls-parsetree-the-first-ocamls-ast). The latter is called "PPX," an
+acronym for Pre-Processor eXtension.
+
+While both types of preprocessing have their use cases, in OCaml it is
+recommended to use PPXs whenever possible for several reasons:
+- They integrate very nicely with Merlin and Dune, and they won't interfere
+  with features such as error reporting in an editor and Merlin's "jump to definition."
+- They are fast and compose well.
```

Comment:
I'll rewrite this with something along the lines of
```suggestion
- They are fast and compose well, when properly written (link to the section on ppxlib).
```

Reply:
> I'll rewrite this with something along the lines of

Please be sure to wrap `ppxlib` in monospace if it will be in the sentence.

Reply:
Cool, sounds good! As a side-note: you could also say "locally" instead of "properly" (a whole-file transformation is also a "proper" PPX in a context where context-free files aren't possible).